MinPlatformPkg/Flash/SpiFvbService/FvbInfo.c

```diff
@@ -11,66 +12,123 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 
 #define FIRMWARE_BLOCK_SIZE         0x10000
 #define FVB_MEDIA_BLOCK_SIZE        FIRMWARE_BLOCK_SIZE
-
-#define NV_STORAGE_BASE_ADDRESS     FixedPcdGet32(PcdFlashNvStorageVariableBase)
-#define SYSTEM_NV_BLOCK_NUM         ((FixedPcdGet32(PcdFlashNvStorageVariableSize)+ FixedPcdGet32(PcdFlashNvStorageFtwWorkingSize) + FixedPcdGet32(PcdFlashNvStorageFtwSpareSize))/ FVB_MEDIA_BLOCK_SIZE)
-
 typedef struct {
   EFI_PHYSICAL_ADDRESS        BaseAddress;
   EFI_FIRMWARE_VOLUME_HEADER  FvbInfo;
   EFI_FV_BLOCK_MAP_ENTRY      End[1];
 } EFI_FVB2_MEDIA_INFO;
 
-//
-// This data structure contains a template of all correct FV headers, which is used to restore
-// Fv header if it's corrupted.
-//
-EFI_FVB2_MEDIA_INFO mPlatformFvbMediaInfo[] = {
-  //
-  // Systen NvStorage FVB
-  //
-  {
-    NV_STORAGE_BASE_ADDRESS,
-    {
-      {0,}, //ZeroVector[16]
-      EFI_SYSTEM_NV_DATA_FV_GUID,
-      FVB_MEDIA_BLOCK_SIZE * SYSTEM_NV_BLOCK_NUM,
-      EFI_FVH_SIGNATURE,
-      0x0004feff, // check MdePkg/Include/Pi/PiFirmwareVolume.h for details on EFI_FVB_ATTRIBUTES_2
-      sizeof (EFI_FIRMWARE_VOLUME_HEADER) + sizeof (EFI_FV_BLOCK_MAP_ENTRY),
-      0,    //CheckSum which will be calucated dynamically.
-      0,    //ExtHeaderOffset
-      {0,}, //Reserved[1]
-      2,    //Revision
-      {
-        {
-          SYSTEM_NV_BLOCK_NUM,
-          FVB_MEDIA_BLOCK_SIZE,
-        }
-      }
-    },
-    {
-      {
-        0,
-        0
-      }
-    }
+// MU_CHANGE - START - TCBZ3478 - Add Dynamic Variable Store and Microcode Support
+/**
+  Returns FVB media information for NV variable storage.
+
+  @return       FvbMediaInfo          A pointer to an instance of FVB media info produced by this function.
+                                      The buffer is allocated internally to this function and it is the caller's
+                                      responsibility to free the memory
+
+**/
+typedef
+EFI_STATUS
+(*FVB_MEDIA_INFO_GENERATOR)(
+  OUT EFI_FVB2_MEDIA_INFO     *FvbMediaInfo
+  );
+
+/**
+  Returns FVB media information for NV variable storage.
+
+  @return       FvbMediaInfo          A pointer to an instance of FVB media info produced by this function.
+                                      The buffer is allocated internally to this function and it is the caller's
+                                      responsibility to free the memory
+
+**/
+EFI_STATUS
+GenerateNvStorageFvbMediaInfo (
+  OUT EFI_FVB2_MEDIA_INFO     *FvbMediaInfo
+  )
+{
+  EFI_STATUS                  Status;
+  UINT32                      NvBlockNum;
+  UINT32                      NvStorageVariableSize;
+  UINT64                      TotalNvVariableStorageSize;
+  EFI_PHYSICAL_ADDRESS        NvStorageBaseAddress;
+  EFI_FIRMWARE_VOLUME_HEADER  FvbInfo = {

```

Comment:
complex stack initializers like this can cause compilers to emit calls to e.g. memset which can cause portability issues.

Reply:
I agree. This is borrowed from the code already there so it shouldn't be introducing a regression. It would be a future improvement to simplify it.